File: Data Structures & Algorithms/partition-equal-subset-sum/submission-3.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        import copy        
        #backtracking. Can keep track of only one subset and can infer the rest

        def dfs(i, s1, s2):

            #end case. Check the end conditions here
            if i == len(nums):
                if s1 == s2:
                    return True
                return False
            
            #not take
            t1 = dfs(i+1, s1, s2 + nums[i])

            #take
            t2 = dfs(i+1, s1  + nums[i], s2)
            
            return t1 or t2

        return dfs(0, 0, 0)

        #lets try to optimize it

File: Data Structures & Algorithms/append-characters-to-string-to-make-subsequence/submission-3.cpp
class Solution {
public:
    int appendCharacters(string s, string t) {
        int sn= s.size(); int tn= t.size();

        int l = 0; int r = 0;

        while (l < sn && r < tn) {
            if (s[l] == t[r]) { r++;}
            l++;
        }

        return tn-r;
        
    }
};